test/extended/networking/vlan.go

```diff
@@ -0,0 +1,92 @@
+	kapiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var _ = g.Describe("[sig-network][Feature:vlan]", func() {
+	oc := exutil.NewCLI("vlan")
+	f := oc.KubeFramework()
+	for _, pluginType := range []string{"vlan", "ipvlan", "macvlan"} {
+
+		vlanNadConfig := `{
+		"cniVersion":"0.4.0","name":"%s",
+		"plugins":[
+			{ 
+				"type": "%s", 
+				"linkInContainer": true,
+				"master": "%s", 
+				"mtu": 1500, 
+				"vlanId": 73,
+				"ipam": { "type": "static", "addresses": [{ "address": "%s" }] }
+           }]
+		}`
+
+		g.It(fmt.Sprintf("should create pingable pods with %s interface on an in-container master [apigroup:k8s.cni.cncf.io]", pluginType), func() {
+			namespace := f.Namespace.Name
+			podName1, podName2 := "pod1", "pod2"
+			vlan1, vlan2 := "vlan1", "vlan2"
+			ip1, ip2 := "10.10.0.2/24", "10.10.0.3/24"
+			bridge := "br0"
+
+			g.By("creating bridge network attachment definition")
+			err := createNetworkAttachmentDefinition(
+				oc.AdminConfig(),
+				namespace,
+				bridge,
+				fmt.Sprintf(`{"cniVersion":"0.4.0","name":"%s","plugins":[{ "type": "bridge", "bridge": "%s"}]}`, bridge, bridge),
+			)
+
+			g.By("creating a vlan network attachment definition")
+			err = createNetworkAttachmentDefinition(
+				oc.AdminConfig(),
+				namespace,
+				vlan1,
+				fmt.Sprintf(vlanNadConfig, vlan1, pluginType, bridge, ip1),
+			)
+			o.Expect(err).NotTo(o.HaveOccurred(), fmt.Sprintf("unable to create %s network-attachment-definition", pluginType))
+
+			err = createNetworkAttachmentDefinition(
+				oc.AdminConfig(),
+				namespace,
+				vlan2,
+				fmt.Sprintf(vlanNadConfig, vlan2, pluginType, bridge, ip2),
+			)
+			o.Expect(err).NotTo(o.HaveOccurred(), fmt.Sprintf("unable to create %s network-attachment-definition", pluginType))
+
+			g.By("creating first pod and checking results")
+			exutil.CreateExecPodOrFail(f.ClientSet, namespace, podName1, func(pod *kapiv1.Pod) {
+				bridgeAnnotation := fmt.Sprintf("%s/%s@%s", namespace, bridge, bridge)
+				vlanAnnotation := fmt.Sprintf("%s/%s@%s", namespace, vlan1, vlan1)
+				pod.ObjectMeta.Annotations = map[string]string{"k8s.v1.cni.cncf.io/networks": fmt.Sprintf("%s, %s", bridgeAnnotation, vlanAnnotation)}
+			})
+			pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(context.TODO(), podName1, metav1.GetOptions{})
+			o.Expect(err).ToNot(o.HaveOccurred())
+
+			networkStatusString, ok := pod.Annotations["k8s.v1.cni.cncf.io/network-status"]
+			o.Expect(ok).To(o.BeTrue())
+			o.Expect(networkStatusString).ToNot(o.BeNil())
+			networkStatuses := []nadtypes.NetworkStatus{}
+			o.Expect(json.Unmarshal([]byte(networkStatusString), &networkStatuses)).ToNot(o.HaveOccurred())
+			o.Expect(networkStatuses).To(o.HaveLen(3))
+			o.Expect(networkStatuses[2].Interface).To(o.Equal(vlan1))
+			o.Expect(networkStatuses[2].Name).To(o.Equal(fmt.Sprintf("%s/%s", namespace, vlan1)))
+
+			g.By("having a second pod pinging the first pod")
```

Comment:
This is the happy path. Can we add a negative test to assert that traffic between pods on different VLAN IDs doesn't work?

Reply:
Done